Add pre-processing flag and functions applicable to fMOST datastes

noisysky opened the pull request:
Add `--pre-processing` command line option, which accepts strings such as 'fmost', 'none', 'default'.
Add pre-processing functions specifically for fMOST datasets, which remove horizontal stripes and background.

adamltyson commented:
Hey @noisysky, thanks for this PR. A couple of things:
* OpenCV can sometimes cause installation issues on different OSs, is there a version of the function you need in numpy, scipy, scikit-image etc?
* Could you add some tests for this new functionality?
* Is there some example data to test this new functionality?

noisysky commented:
@adamltyson Thanks for the review! 
- Replaced OpenCV connectedComponents with skimage label and regionprops.
- Added 1 test (if it's not enough, please help me to come up with ideas for tests).
- The test data can be found here: https://download.brainlib.org/hackathon/2022_GYBS/input/fMOST/subject/. I had to convert .nii.gz files to .tif before running brainreg on them.

adamltyson commented:
Hey @noisysky, sorry for the very delayed response. This looks good, I've left some comments in the review.

BTW - brainreg will accept `.nii` files, just not the compressed versions. I want to extend this, but in the context of a general purpose I/O library (see image.sc discussion - https://forum.image.sc/t/developing-a-common-framework-to-represent-anatomical-data/68860).

noisysky commented on brainreg/cli.py:
```diff
@@ -165,6 +166,23 @@ def geometry_parser(parser):
+        "Pre-processing options"
+    )
+
+    preprocessing_opt_parser.add_argument(
```
Yes, added the default value.

noisysky commented on tests/tests/test_fmost_preprocessing.py:
```diff
@@ -0,0 +1,21 @@
+from brainreg.utils.preprocess import subtract_background
+
+
+def test_background_removal():
```
Yes, I uploaded a pair of images to the test/data/input folder, and the test you suggested runs on them.

noisysky commented on brainreg/backend/niftyreg/run.py:
```diff
@@ -49,13 +50,18 @@ def run_niftyreg(
         niftyreg_paths.downsampled_filtered,
     )
 
+    imio.to_tiff(
```
I removed it. It was there for debug and I forgot about it :)

noisysky commented on brainreg/utils/preprocess.py:
```diff
@@ -1,19 +1,27 @@
+    if preprocessing_args and preprocessing_args.preprocessing == "none":
+        pass
+    elif preprocessing_args and preprocessing_args.preprocessing == "fmost":
+        for i in trange(brain.shape[0], desc="filtering", unit="plane"):  # only coronal plane because of stripes
```
The brain at this point is always reoriented to the atlas orientation. With that comment I tried to justify why I am slicing images in the other dimension than the default pre-processing: in coronal plane stripes are most visible / easiest to remove. I removed the comment.

noisysky commented on brainreg/paths.py:
```diff
@@ -14,6 +14,7 @@ def __init__(self, registration_output_folder):
 
     def make_reg_paths(self):
         self.downsampled_brain_path = self.make_reg_path("downsampled.tiff")
+        self.downsampled_filtered_brain_path = self.make_reg_path("downsampled_filtered.tiff")
```
Removed this as well

noisysky commented on brainreg/utils/preprocess.py:
```diff
@@ -61,3 +69,135 @@ def pseudo_flatfield(img_plane, sigma=5):
+    return fshift
+
+
+def remove_stripes(img_plane, stripes_direction="h"):
```
It looks like all fMOST datasets that I've seen have the same stripes direction, so I only retained "horizontal stripes" part of the code.

adamltyson commented:
Hey @noisysky, this is looking in pretty good shape now (I left a few more comments). A couple more things though:
* It would be great to add some more tests to at least make sure the test coverage % doesn't drop.
* When running brainreg on some fmost datasets, I saw some signal missing in the `downsample_filtered.nii` image that's used for registration. This isn't present in the raw data (or the downsampled stack). There's also a misalignment around the cerebellum, but that's present in the raw data.
![Screenshot from 2022-08-10 13-10-28](https://user-images.githubusercontent.com/13147259/183898077-5d03f795-4f0e-410e-89a2-278082386ca2.png)

Have you seen this before with the preprocessing?

noisysky commented:
Hi @adamltyson, there was a major problem that few z planes got almost entirely deleted. Please give me a few days to check if the new version of the processing that I have performs better on this data.

noisysky commented:
![15257_no_bg_fft_for_PR](https://user-images.githubusercontent.com/33125830/201973044-270d2fce-b209-4720-bd30-fd3ba03b3e14.png)
@adamltyson My newer version doesn't produce as bad artifacts as in your image above, but produces minor artifacts anyway. Not sure if I can improve it much. I abandoned doing background removal using standard methods, and moved on to RandomForest pixel classifiers for this purpose. Also, I noticed that fMOST align better without any pre-processing at all, compared to standard brainreg pre-processing, therefore it makes sense to keep the option of disabling the pre-processing.
I'll push the newest version of background removal that I have, and let you decide whether you want it.

adamltyson commented:
> Also, I noticed that fMOST align better without any pre-processing at all, compared to standard brainreg pre-processing, therefore it makes sense to keep the option of disabling the pre-processing.

In that case, would you mind if I just kept the option of disabling pre-processing, and not the fMOST-specific preprocessing? Long term, I want to move away from the brainreg CLI, and move to:
* Standard - combining brainreg napari plugin with flexible preprocessing provided by other napari plugins
* Advanced users - using brainreg Python API along with custom preprocessing

noisysky commented:
@adamltyson Sure! If I have a radical improvement, I'll let you know and we'll open a new PR.

adamltyson commented:
Great, thanks. At some point I'll move the preprocessing into it's own module, then I'll check back to see what the current status of fmost preprocessing is.
